### PR TITLE
Mln635 delayed job updates user

### PR DESCRIPTION
**Pull request by daryanypl**

### CODE REVIEW

File: app/jobs/find_available_user_barcode_job.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+# Asynchronously calls User methods to find a new barcode that
+# would be available in both MLN db and Sierra.
+class FindAvailableUserBarcodeJob < ApplicationJob
```
Comment (gonuguntla) [inline app/jobs/find_available_user_barcode_job.rb:5]: @daryanypl  Please can you add readme document for this job.

File: app/models/user.rb [hunk 2]
```diff
@@ -408,15 +492,35 @@ def find_unique_new_barcode
...
+       'status' => "start"
+      })
 
-    # TODO: future branches: move user.assign_barcode over to its own multithreaded job,
-    # and move save_as_complete over to after have successfully sent to Sierra
-    self.assign_barcode!
-    self.save_as_complete!
+    # Enqueue a job to be performed as soon as the queuing system is free.
+    begin
+      # Note: user.pin is not getting serialized properly, hence passing as its own var
```
Comment (thisisstephenbetts) [inline app/models/user.rb:503]: Please add into this comment that `pin` is not persisted in the database, so it is not possible to retrieve it later

**Review by thisisstephenbetts — CHANGES_REQUESTED** [review]
Thanks. I've a few comments, but the main issue is the blank pin_code conditional.

File: app/models/user.rb [hunk 3]
```diff
@@ -408,15 +492,35 @@ def find_unique_new_barcode
...
+      FindAvailableUserBarcodeJob.perform_later(user: self, pin_code: self.pin)
+    rescue StandardError => exception
+      LogWrapper.log('ERROR', {
+          'method' => "#{model_name}.find_unique_new_barcode",
+          'message' => "Sierra threw an error while looking for user(#{self.id || 'No ID Available'}) barcode in Sierra: (#{exception.message})"
+        })
+      raise exception
+    rescue Exceptions => exception
```
[the remark this answers is not in the document]
Comment (daryanypl) [reply app/models/user.rb:511]: This functionality will be changed once we upgrate to 5.2.  For now, this is what we need for this PR.  But stay tuned, for future PRs.

File: app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -48,14 +48,23 @@ def create
           set_minimum_password_length
           respond_with resource
         end
-      rescue Net::ReadTimeout
+      rescue Net::ReadTimeout => exception
```
[the remark this answers is not in the document]
Comment (daryanypl) [reply app/controllers/registrations_controller.rb:51]: I'd rather not do that.  In my experience as a maintainer, I've found that handling each kind of exception separately allows me to include specific information in the logs, to better see code origins in stack traces, to be able to binding.pry and step through the code more smoothly.  This is a personal preference, and I can see how you might prefer to opt for shorter code.  However, personal preference doesn't rise to level of "bug, can damage code base", and I don't believe it should hold up a PR. 
As the project's maintainer, I'd rather keep the code the way I would find it manageable.

File: app/controllers/registrations_controller.rb [hunk 5]
```diff
@@ -48,14 +48,23 @@ def create
...
+        LogWrapper.log('ERROR', { 'method' => "RegistrationsController.create",
+          'message' => "Creating new patron threw a StandardError: #{exception.message}, with backtrace: #{exception.backtrace.join('\n')}" })
         set_flash_message :notice, :time_out if is_flashing_format?
         render :template => '/devise/registrations/new'
       end
     else
       render :template => '/devise/registrations/new', :locals => { :error_msg_hash => error_msg_hash }
     end
-  rescue Exceptions::InvalidResponse
+  rescue Exceptions::InvalidResponse => exception
```
[the remark this answers is not in the document]
Comment (daryanypl) [reply app/controllers/registrations_controller.rb:65]: You're right, it bears a look.  However, I see some Exceptions::InvalidResponse statements in the logs.  I will keep it for now, and add a backlog ticket to look further into this.

File: app/jobs/find_available_user_barcode_job.rb [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+
+    # TODO
+    # handle race conditions in case more than one server is processing users
+    # or users are incrementing barcodes independently of each other without
+    # locking table
+  end
+
+
+  rescue_from Exception do |exception|
```
[the remark this answers is not in the document]
Comment (daryanypl) [reply app/jobs/find_available_user_barcode_job.rb:98]: I've been able to trigger it once or twice in testing, yes.  Don't remember offhand when.

File: app/models/user.rb [hunk 7]
```diff
@@ -120,19 +119,34 @@ def validate_pin_pattern
...
+  # In lieu of writing a custom serializer for User, we chose to pass the pin/pin_code
+  # around manually. 
+  def send_request_to_patron_creator_service(pin_code)
+    LogWrapper.log('DEBUG', {
+       'message' => "user.send_request_to_patron_creator_service: start with self=#{self || 'NA'}",
+       'method' => "#{model_name}.send_request_to_patron_creator_service",
+       'status' => "start"
+    })
+
+    if pin_code.blank?
```
[the remark this answers is not in the document]
Comment (daryanypl) [reply app/models/user.rb:134]: We have a backlog jira ticket for this discussion.

File: app/jobs/find_available_user_barcode_job.rb [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    Delayed::Worker.logger.error("#{self.class.name}.rescue_from: threw an error: #{exception.message || 'nil'}")
+  end
+
+
+
+  private
+
+  def around_cleanup
```
[the remark this answers is not in the document]
Comment (daryanypl) [reply app/jobs/find_available_user_barcode_job.rb]: done

File: app/models/user.rb (hunk 2 above)
Comment (daryanypl) [reply app/models/user.rb]: done

File: app/jobs/find_available_user_barcode_job.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (daryanypl) [reply app/jobs/find_available_user_barcode_job.rb:98]: I made a jira ticket for the backlog, to remind ourselves to re-do exception handling, once we upgrade rails.

**Pull request merged**